sherlockml/datasets/__init__.py

```diff
@@ -0,0 +1,583 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import fnmatch
+import os
+import posixpath
+import contextlib
+import tempfile
+import io
+
+from botocore.client import ClientError
+
+from sherlockml.datasets import path, session
+from sherlockml.datasets.session import SherlockMLDatasetsError
+
+
+# For backwards compatability
+SherlockMLFileSystemError = SherlockMLDatasetsError
+
+
+def _s3_client(project_id=None):
+    """Generate a boto S3 client to access this project's datasets."""
+
+    # At present, calls to the interface of this module run this function once
+    # to generate an S3 client, then pass it around to avoid making many calls
+    # to the secret service. This could result in an invalid client in the
+    # middle of an operation, if we get unlucky and the AWS credentials expire
+    # at that moment.
+
+    # In the future, we may wish to make a wrapper class that manages the S3
+    # client and proxies calls to its methods. This wrapper would catch
+    # failures resulting from invalid AWS credentials and build a new client
+    # and retry the operation accordingly.
+
+    return session.get().s3_client(project_id)
+
+
+def _bucket(project_id=None):
+    return session.get().bucket(project_id)
+
+
+def ls(prefix='/', project_id=None, show_hidden=False, s3_client=None):
+    """List contents of project datasets.
+
+    Parameters
+    ----------
+    prefix : str, optional
+        List only files in the datasets matching this prefix. Default behaviour
+        is to list all files.
+    project_id : str, optional
+        The project to list files from. You need to have access to this project
+        for it to work. Defaults to the project set by SHERLOCKML_PROJECT_ID in
+        your environment.
+    show_hidden : bool, optional
+        Include hidden files in the output. Defaults to False.
+    s3_client : botocore.client.S3, optional
+        Advanced - a specific boto client for AWS S3 to use.
+
+    Returns
+    -------
+    list
+        The list of files from the project datasets.
+    """
+
+    if s3_client is None:
+        s3_client = _s3_client(project_id)
+
+    bucket = _bucket(project_id)
+
+    # Use a paginator to enable listing projects with more than 1000 files
+    paginator = s3_client.get_paginator('list_objects_v2')
+    response_iterator = paginator.paginate(
+        Bucket=bucket,
+        Prefix=path.projectpath_to_bucketpath(prefix, project_id)
+    )
+
+    paths = []
+    for part in response_iterator:
+
+        try:
+            objects = part['Contents']
+        except KeyError:
+            continue
+
+        for obj in objects:
+            project_path = path.bucketpath_to_projectpath(obj['Key'])
+
+            # Ignore the root of the project directory
+            if project_path != '/':
+                paths.append(project_path)
+
+    if show_hidden:
+        return paths
+    else:
```

Comment:
I have a slight preference for the current conceptual usage of the if statement to determine the two return cases rather than short circuiting in the first case.